src/main/java/com/microsoft/sqlserver/jdbc/KeyVaultTokenCredential.java

```diff
@@ -0,0 +1,205 @@
+     *
+     * @param authorization The name of the authorization.
+     * @return The updated {@link KeyVaultTokenCredential} instance.
+     */
+    KeyVaultTokenCredential setAuthorization(String authorization) {
+        if (null != this.authorization && this.authorization.equals(authorization)) {
+            return this;
+        }
+        this.authorization = authorization;
+        confidentialClientApplication = getConfidentialClientApplication();
+        return this;
+    }
+
+    /**
+     * Creates an instance of {@link ConfidentialClientApplication} using the provided client id and secret.
+     *
+     * @return An instance of {@link ConfidentialClientApplication}.
+     */
+    private ConfidentialClientApplication getConfidentialClientApplication() {
+        if (null == clientId) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_NullValue"));
+            Object[] msgArgs1 = {"Client ID"};
+            throw new IllegalArgumentException(form.format(msgArgs1), null);
+        }
+
+        if (null == authorization) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_NullValue"));
+            Object[] msgArgs1 = {"Authorization"};
+            throw new IllegalArgumentException(form.format(msgArgs1), null);
+        }
+
+        if (null == clientSecret) {
+            MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_NullValue"));
+            Object[] msgArgs1 = {"Client Secret"};
+            throw new IllegalArgumentException(form.format(msgArgs1), null);
+        }
+
+        // Create the credential using the MSAL factory method.
+        IClientCredential credential;
+        credential = ClientCredentialFactory.createFromSecret(clientSecret);
+        ConfidentialClientApplication.Builder applicationBuilder = ConfidentialClientApplication.builder(clientId,
+                credential);
+        try {
+            applicationBuilder = applicationBuilder.authority(authorization);
+        } catch (MalformedURLException e) {
+            throw logger.logExceptionAsWarning(new IllegalStateException(e));
+        }
+        return applicationBuilder.build();
+    }
+
+    /**
+     * Attempts to get the access token from the client cache if it's not expired. If it's expired this returns an
+     * empty response.
+     * @param request The context for requesting the token including the scope.
+     * @return The cached access token if it's not expired.
+     */
+    private Mono<AccessToken> authenticateWithConfidentialClientCache(TokenRequestContext request) {
+        return Mono.fromFuture(() -> {
+            SilentParameters.SilentParametersBuilder parametersBuilder = SilentParameters
+                    .builder(new HashSet<>(request.getScopes()));
+            try {
+                return confidentialClientApplication.acquireTokenSilently(parametersBuilder.build());
+            } catch (MalformedURLException e) {
```

Comment:
Could you explain why we log a warning here instead of an exception?

Reply:
it doesn't it's actually throwing a RuntimeException and doesn't need to be logged as that' fatal